File: read-spark-backend/internal/domain/article.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type Article struct {
	ID          uuid.UUID  `gorm:"type:uuid;primary_key" json:"id"`
	Title       string     `gorm:"not null;size:500" json:"title"`
	Summary     *string    `json:"summary,omitempty"`
	Content     string     `gorm:"not null;type:text" json:"content"`
	Translation *string    `gorm:"type:text" json:"translation,omitempty"`
	Category    string     `gorm:"not null;size:50" json:"category"`
	Difficulty  string     `gorm:"not null;size:10" json:"difficulty"`
	WordCount   int        `gorm:"default:0" json:"word_count"`
	AudioURL    *string    `gorm:"size:500" json:"audio_url,omitempty"`
	CoverImage  *string    `gorm:"size:500" json:"cover_image,omitempty"`
	IsPremium   bool       `gorm:"default:true" json:"is_premium"`
	PublishedAt *time.Time `json:"published_at,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

type ArticleSummary struct {
	ID          uuid.UUID  `json:"id"`
	Title       string     `json:"title"`
	Summary     *string    `json:"summary,omitempty"`
	Category    string     `json:"category"`
	Difficulty  string     `json:"difficulty"`
	WordCount   int        `json:"word_count"`
	CoverImage  *string    `json:"cover_image,omitempty"`
	IsPremium   bool       `json:"is_premium"`
	PublishedAt *time.Time `json:"published_at,omitempty"`
}

type SearchQuery struct {
	Keyword    string
	Category   *string
	Difficulty *string
	Page       int
	PageSize   int
}

type SearchResult struct {
	Articles []ArticleSummary
	Total    int64
}

// ArticleSearcher interface - allows swapping PG full-text for Elasticsearch later
type ArticleSearcher interface {
	Search(ctx context.Context, query SearchQuery) (SearchResult, error)
	Index(ctx context.Context, article Article) error
	Delete(ctx context.Context, articleID uuid.UUID) error
}

File: read-spark-ios/ReadSpark/Core/Models/Article.swift
import Foundation

struct Article: Codable, Identifiable {
    let id: UUID
    let title: String
    let summary: String?
    let content: String
    let translation: String?
    let category: String
    let difficulty: String
    let wordCount: Int
    let audioUrl: String?
    let coverImage: String?
    let isPremium: Bool
    let publishedAt: String?
}

struct ArticleSummary: Codable, Identifiable {
    let id: UUID
    let title: String
    let summary: String?
    let category: String
    let difficulty: String
    let wordCount: Int
    let coverImage: String?
    let isPremium: Bool
    let publishedAt: String?
}

struct TokenPair: Codable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
}

struct ReadingProgress: Codable, Identifiable {
    let id: UUID
    let articleId: UUID
    let position: Int
    let percentage: Double
    let lastReadAt: String
}
